Inject extension garden access into `ControllerInstallations` (#8204)
gardener/gardener: docs/extensions/garden-api-access.md

# Access to the Garden Cluster for Extensions

> ## TL;DR
> 

Reviewer: ```suggestion
```

Extensions that are installed on seed clusters via a `ControllerInstallation` can simply read the kubeconfig file specified by the `GARDEN_KUBECONFIG` environment variable to create a garden cluster client.

> With this, they use a short-lived token (valid for `12h`) for a dedicated `ServiceAccount` in the `seed-<seed-name>` namespace to securely access the garden cluster.

Reviewer: ```suggestion
With this, they use a short-lived token (valid for `12h`) associated with a dedicated `ServiceAccount` in the `seed-<seed-name>` namespace to securely access the garden cluster.
```

This will instruct gardenlet to create a new `ServiceAccount` named `example` in its own `seed-<seed-name>` namespace in the garden cluster, request a token for it, and populate the token in the secret's data under the `token` key.

## Using Gardenlet-Managed Garden Access

By default, extensions are equipped with secure access to the garden cluster using a dedicated `ServiceAccount` without requiring any additional action.
They can simply read the file specified by the `GARDEN_KUBECONFIG` and construct a garden client with it.

When installing a [`ControllerInstallation`](controllerregistration.md), gardenlet creates two secrets in the installation's namespace: a generic garden kubeconfig (`generic-garden-kubeconfig-<hash>`) and a garden access secret (`garden-access-extension`).

> Additionally, it injects `volume`, `volumeMounts`, and two environment variables into all objects in the `apps` and `batch` API group:

Reviewer: ```suggestion
Additionally, it injects `volume`, `volumeMounts`, and two environment variables into all (init) containers in all resources in the `apps` and `batch` API groups:
```

- The `GARDEN_KUBECONFIG` environment variable points to the path where the generic garden kubeconfig is mounted.

>   This in turn contains the garden cluster's address and current CA bundle as well as the path where the garden access token is mounted.

Reviewer: This sentence is a bit confusing 👀 Maybe omit it or improve?

        - name: SEED_NAME
          value: local
        volumeMounts:
        - mountPath: /var/run/secrets/gardener.cloud/garden/generic-kubeconfig
          name: garden-kubeconfig
          readOnly: true
      volumes:
      - name: garden-kubeconfig
        projected:
          defaultMode: 420
          sources:
          - secret:
              items:
              - key: kubeconfig
                path: kubeconfig
              name: generic-garden-kubeconfig-81fb3a88
              optional: false
          - secret:
              items:
              - key: token
                path: token
              name: garden-access-extension
              optional: false
```

The generic garden kubeconfig will look like this:

```yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority-data: LS0t...
    server: https://garden.local.gardener.cloud:6443
  name: garden
contexts:
- context:
    cluster: garden
    user: extension
  name: garden
current-context: garden
users:
- name: extension
  user:
    tokenFile: /var/run/secrets/gardener.cloud/garden/generic-kubeconfig/token
```

## Using Self-Managed Garden Access

> Extensions can also manage their own garden access secrets if required.

Reviewer: Maybe it makes sense to provide an example for this use case and perhaps example code? It's quite abstract from just reading it here?

Author: Right, this is too abstract to be useful. My goal behind this section was only to document the exceptions of the injection mechanism, i.e., skipping objects that already have the `GARDEN_KUBECONFIG` env var. 
Instead of documenting this in its own section, I merged it with the previous one. Folks who have concrete use cases for this will know what to do :D
